### PR TITLE
Fix JDBC connectors to clean up temp table after exception

### PR DESCRIPTION
**Pull request by tdcmeehan**
JDBC based connectors use the following strategy to insert into the
destination table:

* Create a temp table which has the same columns as the destination
table
* Direct JdbcPageSink to write to this temp table
* Select into the destination table from the temp table

If an exception occurs during the write, then no attempt is made to
clean up this temp table.  This commit adds code to attempt this clean
up during the phases where we could receive errors from the destination
database and which may result in orphaned tables.

### CODE REVIEW

**Comment by findepi** [comment]
When writes fails, the Presto transaction should be rolled back. Shouldn't this be handled by some `setRollback` in `com.facebook.presto.plugin.jdbc.JdbcMetadata#beginInsert`, just as we do in `com.facebook.presto.plugin.jdbc.JdbcMetadata#beginCreateTable`?

File: presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcPageSink.java [hunk 1]
```diff
@@ -197,6 +205,14 @@ public void abort()
         catch (SQLException e) {
             throw new PrestoException(JDBC_ERROR, e);
         }
+        finally {
+            try {
+                jdbcClient.rollbackCreateTable(handle);
```
Comment (kokosing) [inline presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcPageSink.java]: why is this needed if whole transaction is rollbacked, then table should be dropped as well, am I right?

Also `JdbcPageSink` has not created this table, it looks fishy that it drops this.

Comment (tdcmeehan) [reply presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcPageSink.java]: The writes will be rolled back, but not the create table statement.

I think @findepi's suggestion makes sense here, as it accomplishes the same thing as this code but in the correct place.

File: presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java [hunk 2]
```diff
@@ -148,6 +149,27 @@ public void testTableWithNoSupportedColumns()
...
+
+            execute("ALTER TABLE tpch.test_cleanup ADD CHECK (column_a > 0)");
+
+            Exception thrown = null;
+            try {
+                assertUpdate("INSERT INTO tpch.test_cleanup(column_a) VALUES (0)", 1);
+            }
+            catch (Exception e) {
+                thrown = e;
+                assertQuery("SHOW TABLES", "VALUES 'orders', 'test_cleanup'");
```
Comment (electrum) [inline presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java]: The test queries are run in multiple threads when using the Maven test runner, so this will fail due to other tests creating tables. Instead, you can use `SHOW TABLES LIKE 'tmp_presto_%'` and validate that nothing exists.

File: presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcMetadata.java [hunk 3]
```diff
@@ -200,7 +200,9 @@ public void rollback()
     @Override
     public ConnectorInsertTableHandle beginInsert(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
-        return jdbcClient.beginInsertTable(getTableMetadata(session, tableHandle));
+        JdbcOutputTableHandle handle = jdbcClient.beginInsertTable(getTableMetadata(session, tableHandle));
+        setRollback(() -> jdbcClient.rollbackCreateTable(handle));
```
Comment (electrum) [inline presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcMetadata.java]: This needs to be a new `rollbackInsert` method. With this code, we would drop the table we are inserting into (not the temp table).

**Comment by tdcmeehan** [comment]
Squashed and incorporated feedback.  Thank you!

File: presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java [hunk 4]
```diff
@@ -148,6 +149,27 @@ public void testTableWithNoSupportedColumns()
...
+            }
+            assertNotNull(thrown, "Expected exception to be thrown when attempting to insert");
```
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java]: Great tip on `assertQueryFails`, it does exactly what I need here.

File: presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcMetadata.java (hunk 3 above)
Comment (electrum) [reply presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcMetadata.java]: Ah, you are right. I forgot that we actually use a temporary name and rename it at the end.

**Comment by tdcmeehan** [comment]
Squashed and fixed up remaining feedback.  Thank you.

File: presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcMetadata.java (hunk 3 above)
Comment (findepi) [reply presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcMetadata.java]: Can we have  a test that ensures the target table remains alive?

Comment (tdcmeehan) [reply presto-base-jdbc/src/main/java/com/facebook/presto/plugin/jdbc/JdbcMetadata.java]: Tweaked the test I added so that it shows that.

File: presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java [hunk 5]
```diff
@@ -148,6 +148,22 @@ public void testTableWithNoSupportedColumns()
...
+            throws Exception
+    {
+        try (AutoCloseable table = withTable("tpch.test_cleanup", "(x INTEGER)")) {
+            assertQueryReturnsEmptyResult("SHOW TABLES LIKE 'tmp_presto_%'");
+            assertQuery("SHOW TABLES LIKE 'test_cleanup'", "VALUES 'test_cleanup'");
+
+            execute("ALTER TABLE tpch.test_cleanup ADD CHECK (x > 0)");
+
+            assertQueryFails("INSERT INTO tpch.test_cleanup (x) VALUES (0)", "ERROR: new row .* violates check constraint [\\s\\S]*");
+            assertQueryReturnsEmptyResult("SHOW TABLES LIKE 'tmp_presto_%'");
```
[the remark this answers is not in the document]
Comment (tdcmeehan) [reply presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java]: Right, of course.  Updated so tests are performed in a temporary sandbox schema.

File: presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java [hunk 6]
```diff
@@ -148,6 +149,37 @@ public void testTableWithNoSupportedColumns()
         }
     }
 
+    @Test
+    public void testInsertWithFailureDoesntLeaveBehindOrphanedTable()
+            throws Exception
+    {
+        String schemaName = format("tmp_schema_%s", UUID.randomUUID().toString().replaceAll("-", ""));
+        try (
```
Comment (electrum) [inline presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java]: Nit: we always format this with the first item on the first line

File: presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java [hunk 7]
```diff
@@ -148,6 +149,37 @@ public void testTableWithNoSupportedColumns()
...
+                AutoCloseable schema = withSchema(schemaName);
+                AutoCloseable table = withTable(format("%s.test_cleanup", schemaName), "(x INTEGER)")) {
+            assertQuery(format("SELECT table_name FROM information_schema.tables WHERE table_schema = '%s'", schemaName), "VALUES 'test_cleanup'");
+
+            execute(format("ALTER TABLE %s.test_cleanup ADD CHECK (x > 0)", schemaName));
+
+            assertQueryFails(format("INSERT INTO %s.test_cleanup (x) VALUES (0)", schemaName), "ERROR: new row .* violates check constraint [\\s\\S]*");
```
Comment (electrum) [inline presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java]: Why not use `.*` at the end?

Comment (tdcmeehan) [reply presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java]: Unfortunately `.*` won't match newlines

Comment (findepi) [reply presto-postgresql/src/test/java/com/facebook/presto/plugin/postgresql/TestPostgreSqlIntegrationSmokeTest.java]: - i think it makes sense for `com.facebook.presto.tests.QueryAssertions#assertExceptionMessage` to compile pattern with `DOTALL` flag
- you can embed DOTALL flag in the pattern -- `"(?s)ERROR: ... .*"` -- which is arguably more readable, as it directly conveys the intent
